app/src/main/java/at/roteskreuz/stopcorona/model/repositories/ReportingRepository.kt

```diff
@@ -406,6 +422,60 @@ data class PersonalData(
     val tanSuccessfullyRequested: Boolean = false
 )
 
+data class Validity(
+    val rollingStartIntervalNumber: Int,
+    val rollingPeriod: Int?
+) : Comparable<Validity> {
+
+    /**
+     * Order validity by end of the interval. If [rollingPeriod] is not set assume validity for the
+     * whole day.
+     *
+     * This is not formally correct but good enough for all use cases
+     */
+    override fun compareTo(other: Validity): Int {
+        val rollingEndIntervalNumber = rollingStartIntervalNumber + (rollingPeriod ?: 144)
+        val otherRollingEndIntervalNumber =
+            with(other) { rollingStartIntervalNumber + (rollingPeriod ?: 144) }
+
+        return rollingEndIntervalNumber.compareTo(otherRollingEndIntervalNumber)
+    }
+
+    /**
+     * Special equals operator which ignores the [rollingPeriod] if it is not available on either
+     * object.
+     *
+     * This is required for old keys in the database where the rolling period was not stored
+     */
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is Validity) return false
+
+        if (rollingStartIntervalNumber != other.rollingStartIntervalNumber) return false
+        if (rollingPeriod == other.rollingPeriod) return true
+        // For old keys without a rolling period, ignore the rollingPeriod
+        if (rollingPeriod == null || other.rollingPeriod == null) return true
+
+        return false
+    }
+
+    /**
+     * Special hashCode which ignores the [rollingPeriod]. This leads to more collisions but is
+     * required due to the special [equals] operator.
+     *
+     * This is required for old keys in the database where the rolling period was not stored
+     */
+    override fun hashCode(): Int {
+        return rollingStartIntervalNumber
```

Comment:
We could reduce the collisions a bit further if we would use `rollingStartIntervalNumber + (rollingPeriod ?: 0)`

Reply:
The intention is to have all possible tuples (65346, *) match (65346, null). So they must have the same hash code.